portfolio/src/main/java/com/google/sps/servlets/ThumbsDownServlet.java

```diff
@@ -31,44 +31,82 @@
 import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 
 @WebServlet("/thumbsdown-data")
 public final class ThumbsDownServlet extends HttpServlet {
 
     @Override
-    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {        
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+        if (!userService.isUserLoggedIn()) {
+            response.sendRedirect("/contact.html");
+            return;
+        }
+
+        Entity userInfoEntity = getUserInfoEntity();
+
         // Get comment's id (which was passed as a parameter).
         long id = Long.parseLong(request.getParameter("id"));
 
+        Entity commentEntity = getCommentEntity(id);
+        if (commentEntity == null) {
+            response.setContentType("text/html;");
+            response.getWriter().println("Unable to get comment.");
+            return;
+        }
+        
+        if (!unlikedComment(userInfoEntity, commentEntity)) {
+            changePopularity(commentEntity, 1);
+            addToUnlikedComments(userInfoEntity, commentEntity);
+        } else {
+            changePopularity(commentEntity, -1);
+            removeFromUnlikedComments(userInfoEntity, commentEntity);
+        }
+
+
+        response.sendRedirect("/contact.html");
+        return;
+    }
+
+    // Accesses the datastore to get the UserInfo entity. Returns the entity or null if one does not exist.
+    private Entity getCommentEntity(long id) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
         // Using the id, get the comment's key.
         Key commentEntityKey;
         try {
             commentEntityKey = KeyFactory.createKey("Comment", id);
         } catch(Exception e) {
-            response.setContentType("text/html;");
-            response.getWriter().println("Unable to get comment's key.");
-            return;
+            return null;
         }
 
-        // Instantiate datastore
-        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-
         // Get the comment entity using its key.
         Entity commentEntity;
         try {
             commentEntity = datastore.get(commentEntityKey);
         } catch(Exception e) {
-            response.setContentType("text/html;");
-            response.getWriter().println("Unable to get comment.");
-            return;
+            return null;
         }
 
+        return commentEntity;
+    }
+
+    private void changePopularity(Entity commentEntity, long value) {
```

Comment:
Is there ever a scenario where you change the popularity by a number other than 1?

If not, I'd suggest two separate methods incrementPopularity and decrementPopularity

You also need to make these methods thread-safe. To do this, you can use a transaction:
https://cloud.google.com/datastore/docs/concepts/transactions#datastore-datastore-transactional-update-java

We can chat more tomorrow about this

Reply:
I just changed the functions to decrement and increment. Since I will be adding transactions to other functions as well, I will be creating a new PR that will be handling that for all the other functions as well.